Serve asynchronously compiled inline assets on error pages

chrisfrank opened the pull request:
### Description of Change
On error pages, serve inline styles and images instead of linking to external files.

### Related Issue
Fixes #3.

### Motivation and Context
We experimented with a synchronous implementation of this in #29, but it complicates the build process. This implementation inlines assets asynchronously on the first error page request, and on subsequent requests serves the compiled assets from memory.

### Checklist

- [x] Ran `npm run lint` and updated code style accordingly
- [x] `npm run test` passes
- [x] PR has a description and all contributors/stakeholder are noted/cc'ed
- [x] tests are updated and/or added to cover new code
- [x] relevant documentation is changed and/or added

chrisfrank commented on server/routes/errors.js:
```diff
@@ -1,7 +1,52 @@
+    new Promise((resolve, reject) => {
+      fs.readFile('public/errors.css', (err, data) => {
+        if (err) throw err
+        assets.css = data.toString()
```
@isaacwhite if we still want to parse `background-image: url(/etc.png)` from the CSS and inline the images, I think this is where I would add it. I didn't end up needing any background images in this implementation, so I skipped this step for now.

isaacwhite replied:
That makes sense, @chrisfrank. What do you think about leaving an inline comment for future PRs here in case someone else wants to add support for that later?

isaacwhite commented on public/errors.css:
```diff
@@ -0,0 +1,755 @@
+@import url("https://fonts.googleapis.com/css?family=Alfa+Slab+One|Open+Sans:400,400i,700,700i&subset=cyrillic,cyrillic-ext,greek,greek-ext,latin-ext,vietnamese");
```
@chrisfrank we usually don't check in compiled CSS. Since the build task covers both inline and main stylesheets, could we remove this from the diff?

chrisfrank commented:
Hey @isaacwhite,

As per your suggestions, I've removed `public/errors.css` from source control and converted all Promise-based returns to async/await. Turns out removing the compiled stylesheet meant I had to adjust the build process for Travis — which may be why I checked it in to begin with? 🤷‍♂ I can't remember what I was thinking, but I definitely agree that the codebase is cleaner without compiled CSS.

Is there anything else you'd like me to address? I should be able to make time this week.

chrisfrank commented on server/utils.js:
```diff
@@ -101,3 +103,17 @@ exports.stringTemplate = (configPath, ...args) => {
   return ''
 }
+
+const readFileAsync = promisify(fs.readFile)
```
I went with `promisify` because I don't think the `fs.promises` API is supported in Node 8.

isaacwhite replied:
Ah, that's a good catch. Let's reassess at Node 8 EOL in several months, but for now this seems like a good way to go.

chrisfrank commented on server/utils.js:
```diff
@@ -101,3 +103,17 @@ exports.stringTemplate = (configPath, ...args) => {
+  // If the path starts with `/assets`, look in the app’s public directory
+  const publicPath = filePath.replace(/^\/assets/, '/public')
+
+  const mimeType = mime.lookup(path.posix.basename(publicPath))
```
Good questions! My understanding of [the node docs](https://nodejs.org/api/path.html#path_windows_vs_posix) is that because we're using URL-style (and thus POSIX-style) paths here, we need to call `path.posix.basename` to get consistent behavior in both OS families.

> To achieve consistent results when working with POSIX file paths on any operating system, use path.posix

But I'm not set up to test this on Windows, and I may well be misreading the docs. Let me know if you'd like to drop to just `path.basename`.

isaacwhite replied:
Ah, that makes sense. I think you are reading the docs correctly and we should leave this line with the use of `path.posix` as you have done so far. 

Would it make sense to add a 1 line comment explaining the rationale, since it might also strike others as a potential source of bugs? (We don't specify `path.posix` anywhere else in the codebase, which is why this one jumped out at me)

chrisfrank replied:
Good idea — will add that comment.

chrisfrank commented:
Almost there, @isaacwhite! I'll coordinate with @afischer about fallback styles for the search page. It looks like that will happen any time the JS fails to load, not just on error pages, so I think we should handle it in a way that isn't isolated to the `errors.css` file.

chrisfrank commented:
It looks like @afischer added fallback search styles in #65 👍 💯. So I'll pull those changes in, double-check that everything looks like it should, add a comment re: `path.posix.basename`, and then we should be good to go. Thanks for steering this PR along smoothly, @isaacwhite.

chrisfrank commented:
@isaacwhite I think I hit everything on our list today. How's this looking to you? 

FWIW, I don't _think_ the CI failure under Node 10 is related to anything in this PR. When I run the test suite locally from master, I see that same playlist test failing ~10% of the time. I suspect just triggering a new build will make everything pass.

isaacwhite commented:
@chrisfrank this looks great, thanks for taking the time to polish! Excited to add this as core functionality.
